src/metrics/test/metrics_test.cpp

```diff
@@ -502,5 +508,109 @@ TEST_CASE("Registry counts the number of workers") {
     REQUIRE(metrics.getWorkerCount("actor2", "op1") == 2);
 }
 
+TEST_CASE("Events stream to gRPC") {
+    using EventVec = std::vector<poplar::EventMetrics>;
+
+    auto compareEventsAndClear = [](const EventVec& events_in) {
+        internals::v2::MockStreamInterface interface;
+        REQUIRE(events_in.size() == interface.events.size());
+        for (int i = 0; i < events_in.size(); i++) {
+            REQUIRE(google::protobuf::util::MessageDifferencer::Equals(events_in[i], interface.events[i]));
+            //REQUIRE(2 == 1);
+        };
+        interface.events.clear();
+    };
+
+    SECTION("Empty stream") {
+        RegistryClockSourceStub::reset();
+        auto stream = internals::v2::EventStream<RegistryClockSourceStub, 
+             internals::v2::MockStreamInterface>{1, "TestName", 1, "/test/prefix"};
+
+        EventVec expected;
+        compareEventsAndClear(expected);
+    }
+
+    SECTION("Two events") {
```

Comment:
This test is probably pretty brittle but it found some errors that I'd had so there's some value. If necessary, we could template-ize the Registry on the grpc stub to do tests involving it.

Reply:
Riffing: I'm not sure we need to templatize on the whole registry, but we can certainly re-think some of the design assumptions knowing that we can freely use virtuals (we were *very* cautious of *anything* that somebody might be able to "blame" performance anomalies on). I can imagine having OperationImpl be a pure-virtual class (interface) and similar for the grpc glue such that you could easily assert the interactions between Registry<->Operation<->GRPC. That's basically a rewrite of the API but most of the internals stay the same and the testability (probably) goes way up. This is probably *remarkably* easier once we kill csv stuff, though.

ANYWAY - I think you've templatized at a fine level of granularity, but I really think I want to spend a few days cleaning up when we have the csv junk hauled away.